File: interop/php/src/Flow.php
<?php

namespace Flow;

use FFI;
use Exception;

/**
 * Flow Language - PHP Bindings
 * 
 * Allows calling Flow functions from PHP code using FFI.
 */
class Flow
{
    const VERSION = '0.1.0';
    
    private static ?FFI $ffi = null;
    private static bool $initialized = false;

    /**
     * Initialize Flow FFI bindings
     */
    private static function init(): void
    {
        if (self::$initialized) {
            return;
        }

        // Find the library
        $libPath = self::findLibrary();
        
        // Load FFI definitions
        $cdef = <<<'CDEF'
        // Flow value types
        typedef enum {
            FLOW_VAL_INT = 0,
            FLOW_VAL_FLOAT = 1,
            FLOW_VAL_STRING = 2,
            FLOW_VAL_BOOL = 3,
            FLOW_VAL_VOID = 4
        } flow_type_t;

        // Flow value structure
        typedef struct {
            int type;
            union {
                int64_t int_val;
                double float_val;
                const char* string_val;
                bool bool_val;
            } data;
        } flow_value_t;

        // Opaque types
        typedef struct flow_module flow_module_t;

        // Runtime management
        int flow_init(void);
        void flow_cleanup(void);

        // Module management
        flow_module_t* flow_load_module(const char* path);
        flow_module_t* flow_compile_string(const char* source);
        void flow_unload_module(flow_module_t* module);

        // Function calling
        flow_value_t flowc_call_v(flow_module_t* module, const char* function, int argc, flow_value_t* argv);

        // Value constructors
        flow_value_t flow_int(int64_t value);
        flow_value_t flow_float(double value);
        flow_value_t flow_string(const char* value);
        flow_value_t flow_bool(bool value);
        flow_value_t flow_void(void);

        // Value extractors
        int64_t flow_as_int(flow_value_t value);
        double flow_as_float(flow_value_t value);
        const char* flow_as_string(flow_value_t value);
        bool flow_as_bool(flow_value_t value);

        // Error handling
        const char* flow_get_error(void);
        void flow_clear_error(void);

        // Reflection API
        typedef struct {
            const char* name;
            const char* type;
        } flow_param_info_t;

        typedef struct {
            const char* name;
            const char* return_type;
            int param_count;
            flow_param_info_t* params;
        } flow_function_info_t;

        int flow_reflect_function_count(flow_module_t* module);
        int flow_reflect_list_functions(flow_module_t* module, char*** names_out);
        void flow_reflect_free_names(char** names, int count);
        const char* flow_reflect_function_name_at(flow_module_t* module, int index);
        flow_function_info_t* flow_reflect_get_function_info(flow_module_t* module, const char* function_name);
        void flow_reflect_free_function_info(flow_function_info_t* info);
        
        // Bidirectional reflection
        int flow_reflect_register_foreign_module(const char* adapter, const char* module_name, const char** functions, int count);
        int flow_reflect_has_foreign_module(const char* adapter, const char* module_name);
        int flow_reflect_foreign_function_count(const char* adapter, const char* module_name);
        int flow_reflect_foreign_functions(const char* adapter, const char* module_name, char*** names_out);
        CDEF;

        try {
            self::$ffi = FFI::cdef($cdef, $libPath);
        } catch (\FFI\Exception $e) {
            throw new FlowException("Failed to load Flow library: " . $e->getMessage());
        }

        // Initialize runtime
        $ret = self::$ffi->flow_init();
        if ($ret !== 0) {
            $error = self::$ffi->flow_get_error();
            $errorMsg = $error ? FFI::string($error) : "Unknown error";
            self::$ffi->flow_clear_error();
            throw new FlowInitializationException("Failed to initialize Flow runtime: $errorMsg");
        }

        self::$initialized = true;
    }

    /**
     * Find the Flow C library
     */
    private static function findLibrary(): string
    {
        $libName = PHP_OS_FAMILY === 'Darwin' ? 'libflow.dylib' : 'libflow.so';
        
        // Calculate path relative to this file
        // This file is at interop/php/src/Flow.php
        // C library is at interop/c/libflow.dylib
        $thisFile = realpath(__FILE__);
        $interopDir = dirname(dirname(dirname($thisFile)));
        $cLibPath = $interopDir . '/c/' . $libName;
        
        $searchPaths = [
            $cLibPath,                      // Development: interop/c/
            '/usr/local/lib/' . $libName,   // System install
            '/usr/lib/' . $libName,         // System install
        ];

        foreach ($searchPaths as $path) {
            if (file_exists($path)) {
                return $path;
            }
        }

        throw new FlowException(
            "Could not find $libName. Tried:\n" .
            implode("\n", array_map(fn($p) => "  - $p", $searchPaths)) .
            "\n\nPlease build the C library first (see interop/c/README.md)"
        );
    }

    /**
     * Get FFI instance
     */
    public static function getFFI(): FFI
    {
        if (!self::$initialized) {
            self::init();
        }
        return self::$ffi;
    }

    /**
     * Load a Flow module from a file
     * 
     * @param string $path Path to the .flow file
     * @return FlowModule Module with callable functions
     * 
     * @throws FlowCompileException If compilation fails
     */
    public static function loadModule(string $path): FlowModule
    {
        if (!file_exists($path)) {
            throw new \InvalidArgumentException("File not found: $path");
        }

        self::init();
        $runtime = new Runtime();
        
        $handle = self::$ffi->flow_load_module($path);
        if ($handle === null || FFI::isNull($handle)) {
            $error = self::$ffi->flow_get_error();
            $errorMsg = $error ? FFI::string($error) : "Unknown error";
            self::$ffi->flow_clear_error();
            throw new FlowCompileException("Failed to load module: $errorMsg");
        }

        return new FlowModule($handle, $path, $runtime);
    }

    /**
     * Compile Flow source code from a string
     * 
     * @param string $source Flow source code
     * @return FlowModule Module with callable functions
     * 
     * @throws FlowCompileException If compilation fails
     */
    public static function compile(string $source): FlowModule
    {
        self::init();
        $runtime = new Runtime();
        
        $handle = self::$ffi->flow_compile_string($source);
        if ($handle === null || FFI::isNull($handle)) {
            $error = self::$ffi->flow_get_error();
            $errorMsg = $error ? FFI::string($error) : "Unknown error";
            self::$ffi->flow_clear_error();
            throw new FlowCompileException("Failed to compile: $errorMsg");
        }

        return new FlowModule($handle, '<compiled>', $runtime);
    }

    /**
     * Get Flow language version
     * 
     * @return string Version string
     */
    public static function version(): string
    {
        return self::VERSION;
    }

    /**
     * Cleanup on shutdown
     */
    public static function cleanup(): void
    {
        if (self::$initialized && self::$ffi !== null) {
            self::$ffi->flow_cleanup();
            self::$initialized = false;
        }
    }
}

// Register shutdown function
register_shutdown_function([Flow::class, 'cleanup']);

/**
 * Flow runtime manager
 */
class Runtime
{
    private bool $initialized = true;

    public function __construct()
    {
        // Runtime is already initialized by Flow::init()
    }

    public function isInitialized(): bool
    {
        return $this->initialized;
    }
}

/**
 * Represents a compiled Flow module
 */
class FlowModule
{
    private $handle;  // FFI\CData
    private string $path;
    private Runtime $runtime;

    public function __construct($handle, string $path, Runtime $runtime)
    {
        $this->handle = $handle;
        $this->path = $path;
        $this->runtime = $runtime;
    }

    /**
     * Call a Flow function
     * 
     * @param string $name Function name
     * @param array $arguments Function arguments
     * @return mixed Return value
     * 
     * @throws FlowRuntimeException If function call fails
     */
    public function call(string $name, ...$arguments)
    {
        $ffi = Flow::getFFI();
        
        // Convert PHP values to Flow values
        $flowArgs = [];
        foreach ($arguments as $arg) {
            $flowArgs[] = self::phpToFlow($arg);
        }

        // Create C array
        if (!empty($flowArgs)) {
            $cArray = $ffi->new('flow_value_t[' . count($flowArgs) . ']');
            foreach ($flowArgs as $i => $value) {
                $cArray[$i] = $value;
            }
        } else {
            $cArray = null;
        }

        // Call the function
        $result = $ffi->flowc_call_v($this->handle, $name, count($flowArgs), $cArray);

        // Check for errors
        $error = $ffi->flow_get_error();
        if ($error !== null && $error !== '') {
            $ffi->flow_clear_error();
            throw new FlowRuntimeException($error);
        }

        // Convert result back to PHP
        return self::flowToPhp($result);
    }

    /**
     * Magic method to call Flow functions as methods
     */
    public function __call(string $name, array $arguments)
    {
        return $this->call($name, ...$arguments);
    }

    /**
     * Convert PHP value to Flow value
     */
    private static function phpToFlow($value)
    {
        $ffi = Flow::getFFI();

        if (is_bool($value)) {
            return $ffi->flow_bool($value);
        } elseif (is_int($value)) {
            return $ffi->flow_int($value);
        } elseif (is_float($value)) {
            return $ffi->flow_float($value);
        } elseif (is_string($value)) {
            return $ffi->flow_string($value);
        } elseif ($value === null) {
            return $ffi->flow_void();
        } else {
            throw new \InvalidArgumentException("Unsupported type for Flow: " . gettype($value));
        }
    }

    /**
     * Convert Flow value to PHP value
     */
    private static function flowToPhp($value)
    {
        $ffi = Flow::getFFI();
        
        $type = $value->type;
        
        if ($type === 0) { // FLOW_VAL_INT
            return $ffi->flow_as_int($value);
        } elseif ($type === 1) { // FLOW_VAL_FLOAT
            return $ffi->flow_as_float($value);
        } elseif ($type === 2) { // FLOW_VAL_STRING
            $str = $ffi->flow_as_string($value);
            return $str ? FFI::string($str) : null;
        } elseif ($type === 3) { // FLOW_VAL_BOOL
            return $ffi->flow_as_bool($value);
        } elseif ($type === 4) { // FLOW_VAL_VOID
            return null;
        } else {
            throw new FlowRuntimeException("Unknown Flow type: $type");
        }
    }

    public function getPath(): string
    {
        return $this->path;
    }

    // ========================================================================
    // REFLECTION API
    // ========================================================================

    /**
     * Get the number of functions in this module
     * @return int Number of functions
     */
    public function functionCount(): int
    {
        $ffi = Flow::getFFI();
        return $ffi->flow_reflect_function_count($this->handle);
    }

    /**
     * List all function names in this module
     * @return array Array of function names
     */
    public function listFunctions(): array
    {
        $ffi = Flow::getFFI();
        $count = $ffi->flow_reflect_function_count($this->handle);
        
        if ($count === 0) {
            return [];
        }

        // Use simpler approach: get function names one at a time
        $functionNames = [];
        for ($i = 0; $i < $count; $i++) {
            $name = $ffi->flow_reflect_function_name_at($this->handle, $i);
            if ($name !== null && $name !== '') {
                // PHP FFI automatically converts const char* to string
                $functionNames[] = is_string($name) ? $name : FFI::string($name);
            }
        }

        return $functionNames;
    }

    /**
     * Get detailed information about a function
     * @param string $functionName Name of the function
     * @return array Function info with name, return_type, and parameters
     * @throws FlowException If function not found
     */
    public function functionInfo(string $functionName): array
    {
        $ffi = Flow::getFFI();
        $infoPtr = $ffi->flow_reflect_get_function_info($this->handle, $functionName);
        
        if ($infoPtr === null || FFI::isNull($infoPtr)) {
            throw new FlowException("Function '$functionName' not found in module");
        }

        try {
            $info = $infoPtr;
            
            // PHP FFI automatically converts const char* to string
            $result = [
                'name' => is_string($info->name) ? $info->name : FFI::string($info->name),
                'return_type' => is_string($info->return_type) ? $info->return_type : FFI::string($info->return_type),
                'parameters' => []
            ];

            // Extract parameters
            if ($info->param_count > 0 && $info->params !== null && !FFI::isNull($info->params)) {
                for ($i = 0; $i < $info->param_count; $i++) {
                    $param = $info->params[$i];
                    $result['parameters'][] = [
                        'name' => is_string($param->name) ? $param->name : FFI::string($param->name),
                        'type' => is_string($param->type) ? $param->type : FFI::string($param->type)
                    ];
                }
            }

            return $result;
        } finally {
            // Free the function info
            $ffi->flow_reflect_free_function_info($infoPtr);
        }
    }

    /**
     * Get a human-readable string representation of all functions in the module
     * @return string Formatted string with function signatures
     */
    public function inspect(): string
    {
        $functions = $this->listFunctions();
        
        if (empty($functions)) {
            return 'Module contains no functions';
        }

        $lines = ["Module contains " . count($functions) . " function(s):\n"];
        
        foreach ($functions as $funcName) {
            try {
                $info = $this->functionInfo($funcName);
                $params = array_map(
                    fn($p) => "{$p['name']}: {$p['type']}",
                    $info['parameters']
                );
                $lines[] = "  {$info['name']}(" . implode(', ', $params) . ") -> {$info['return_type']}";
            } catch (FlowException $e) {
                $lines[] = "  $funcName (error getting info: {$e->getMessage()})";
            }
        }

        return implode("\n", $lines);
    }

    public function __destruct()
    {
        if ($this->handle !== null) {
            $ffi = Flow::getFFI();
            $ffi->flow_unload_module($this->handle);
        }
    }
}

// Exception classes

class FlowException extends Exception {}
class FlowInitializationException extends FlowException {}
class FlowCompileException extends FlowException {}
class FlowRuntimeException extends FlowException {}

File: interop/ruby/lib/flow.rb
# frozen_string_literal: true

# Flow Language - Ruby Bindings
#
# Ruby bindings for calling Flow functions from Ruby code.
#
# @example
#   require 'flow'
#
#   runtime = Flow::Runtime.new
#   mod = Flow::Module.compile(runtime, <<~FLOW)
#     func add(a: int, b: int) -> int {
#       return a + b;
#     }
#   FLOW
#
#   result = mod.call('add', 10, 20)
#   puts "Result: #{result}"  # => 30

require 'ffi'
require 'tempfile'

module Flow
  VERSION = '0.1.0'

  # FFI bindings to the C library
  module LibFlow
    extend FFI::Library
    
    # Find the library - try multiple locations
    lib_paths = [
      File.expand_path('../../c/libflow.dylib', __dir__),  # macOS dev
      File.expand_path('../../c/libflow.so', __dir__),      # Linux dev
      'libflow.dylib',                                       # macOS installed
      'libflow.so',                                          # Linux installed
      '/usr/local/lib/libflow.dylib',
      '/usr/local/lib/libflow.so'
    ]
    
    lib_path = lib_paths.find { |path| File.exist?(path) }
    raise "Could not find libflow library. Tried: #{lib_paths.join(', ')}" unless lib_path
    
    ffi_lib lib_path
    
    # Value types enum
    enum :flow_type, [
      :int, 0,
      :float, 1,
      :string, 2,
      :bool, 3,
      :void, 4
    ]
    
    # Value union
    class FlowValueData < FFI::Union
      layout :int_val, :int64,
             :float_val, :double,
             :string_val, :pointer,
             :bool_val, :bool
    end
    
    # Value struct
    class FlowValueT < FFI::Struct
      layout :type, :flow_type,
             :data, FlowValueData
    end
    
    # Runtime functions
    attach_function :flow_init, [], :int
    attach_function :flow_cleanup, [], :void
    
    # Module functions
    attach_function :flow_load_module, [:string], :pointer
    attach_function :flow_compile_string, [:string], :pointer
    attach_function :flow_unload_module, [:pointer], :void
    
    # Function calls
    attach_function :flowc_call_v, [:pointer, :string, :int, :pointer], FlowValueT.by_value
    
    # Value constructors
    attach_function :flow_int, [:int64], FlowValueT.by_value
    attach_function :flow_float, [:double], FlowValueT.by_value
    attach_function :flow_string, [:string], FlowValueT.by_value
    attach_function :flow_bool, [:bool], FlowValueT.by_value
    attach_function :flow_void, [], FlowValueT.by_value
    
    # Value extractors
    attach_function :flow_as_int, [FlowValueT.by_value], :int64
    attach_function :flow_as_float, [FlowValueT.by_value], :double
    attach_function :flow_as_string, [FlowValueT.by_value], :string
    attach_function :flow_as_bool, [FlowValueT.by_value], :bool
    
    # Error handling
    attach_function :flow_get_error, [], :string
    attach_function :flow_clear_error, [], :void
    
    # ============================================================
    # REFLECTION API
    # ============================================================
    
    # Parameter info structure
    class FlowParamInfo < FFI::Struct
      layout :name, :pointer,
             :type, :pointer
    end
    
    # Function info structure
    class FlowFunctionInfo < FFI::Struct
      layout :name, :pointer,
             :return_type, :pointer,
             :param_count, :int,
             :params, :pointer  # FlowParamInfo*
    end
    
    # Module reflection
    attach_function :flow_reflect_function_count, [:pointer], :int
    attach_function :flow_reflect_list_functions, [:pointer, :pointer], :int
    attach_function :flow_reflect_free_names, [:pointer, :int], :void
    attach_function :flow_reflect_function_name_at, [:pointer, :int], :string
    
    # Function reflection
    attach_function :flow_reflect_get_function_info, [:pointer, :string], :pointer
    attach_function :flow_reflect_free_function_info, [:pointer], :void
    
    # Bidirectional reflection
    attach_function :flow_reflect_register_foreign_module, [:string, :string, :pointer, :int], :int
    attach_function :flow_reflect_has_foreign_module, [:string, :string], :int
    attach_function :flow_reflect_foreign_function_count, [:string, :string], :int
    attach_function :flow_reflect_foreign_functions, [:string, :string, :pointer], :int
  end

  # Flow runtime error
  class FlowError < StandardError; end
  class CompileError < FlowError; end
  class RuntimeError < FlowError; end
  class InitializationError < FlowError; end

  # Flow runtime manager
  class Runtime
    def initialize
      ret = LibFlow.flow_init
      if ret != 0
        error = LibFlow.flow_get_error
        raise InitializationError, "Failed to initialize Flow runtime: #{error}"
      end
      
      @initialized = true
      ObjectSpace.define_finalizer(self, self.class.finalize)
    end

    def self.finalize
      proc { LibFlow.flow_cleanup }
    end

    def initialized?
      @initialized
    end
  end

  # Represents a compiled Flow module
  class Module
    attr_reader :path

    def initialize(handle, path)
      @handle = handle
      @path = path
      @functions = {}
      
      raise CompileError, "Failed to create module" if @handle.null?
      
      ObjectSpace.define_finalizer(self, self.class.finalize(@handle))
    end

    def self.finalize(handle)
      proc { LibFlow.flow_unload_module(handle) unless handle.null? }
    end

    # Load a Flow module from a file
    #
    # @param runtime [Runtime] Flow runtime instance
    # @param path [String] Path to the .flow file
    # @return [Module] Compiled module
    #
    # @example
    #   runtime = Flow::Runtime.new
    #   mod = Flow::Module.load(runtime, 'math.flow')
    def self.load(runtime, path)
      raise ArgumentError, "File not found: #{path}" unless File.exist?(path)
      
      handle = LibFlow.flow_load_module(path)
      if handle.null?
        error = LibFlow.flow_get_error
        LibFlow.flow_clear_error # Clear error after retrieving
        raise CompileError, "Failed to load module: #{error}"
      end
      
      new(handle, path)
    end

    # Compile Flow source code from a string
    #
    # @param runtime [Runtime] Flow runtime instance
    # @param source [String] Flow source code
    # @return [Module] Compiled module
    #
    # @example
    #   runtime = Flow::Runtime.new
    #   mod = Flow::Module.compile(runtime, <<~FLOW)
    #     func add(a: int, b: int) -> int {
    #       return a + b;
    #     }
    #   FLOW
    def self.compile(runtime, source)
      handle = LibFlow.flow_compile_string(source)
      if handle.null?
        error = LibFlow.flow_get_error
        LibFlow.flow_clear_error # Clear error after retrieving
        raise CompileError, "Failed to compile: #{error}"
      end
      
      new(handle, '<compiled>')
    end

    # Call a Flow function
    #
    # @param function_name [String] Name of the function
    # @param args [Array] Arguments to pass
    # @return [Object] Return value (Integer, Float, String, Boolean, or nil)
    #
    # @example
    #   result = mod.call('add', 10, 20)  # => 30
    def call(function_name, *args)
      # Convert Ruby values to C values
      c_values = args.map { |arg| ruby_to_flow(arg) }
      
      # Create array for FFI
      c_array = FFI::MemoryPointer.new(LibFlow::FlowValueT, c_values.size)
      c_values.each_with_index do |val, i|
        LibFlow::FlowValueT.new(c_array + i * LibFlow::FlowValueT.size).tap do |struct|
          struct[:type] = val[:type]
          struct[:data] = val[:data]
        end
      end
      
      # Call the function
      result = LibFlow.flowc_call_v(@handle, function_name, c_values.size, c_array)
      
      # Check for errors
      error = LibFlow.flow_get_error
      if error && !error.empty?
        LibFlow.flow_clear_error # Clear error after retrieving
        raise RuntimeError, error
      end
      
      # Convert result back to Ruby
      flow_to_ruby(result)
    end

    # Allow calling functions as methods
    #
    # @example
    #   mod.add(10, 20)  # Same as mod.call('add', 10, 20)
    def method_missing(method, *args)
      call(method.to_s, *args)
    rescue RuntimeError => e
      if e.message.include?('Function not found')
        super
      else
        raise
      end
    end

    def respond_to_missing?(method, include_private = false)
      true
    end
    
    # ============================================================
    # REFLECTION API
    # ============================================================
    
    # Get the number of functions in this module
    #
    # @return [Integer] Number of functions
    #
    # @example
    #   count = mod.function_count
    #   puts "Module has #{count} functions"
    def function_count
      raise FlowError, "Module is not loaded" if @handle.null?
      LibFlow.flow_reflect_function_count(@handle)
    end
    
    # List all function names in this module
    #
    # @return [Array<String>] Array of function names
    #
    # @example
    #   functions = mod.list_functions
    #   puts "Functions: #{functions.join(', ')}"
    def list_functions
      raise FlowError, "Module is not loaded" if @handle.null?
      
      count = LibFlow.flow_reflect_function_count(@handle)
      return [] if count == 0
      
      # Allocate pointer to pointer array
      names_ptr = FFI::MemoryPointer.new(:pointer)
      actual_count = LibFlow.flow_reflect_list_functions(@handle, names_ptr)
      
      return [] if actual_count <= 0
      
      # Read the array of string pointers
      names_array = names_ptr.read_pointer
      function_names = []
      
      actual_count.times do |i|
        str_ptr = names_array.get_pointer(i * FFI::Pointer.size)
        function_names << str_ptr.read_string unless str_ptr.null?
      end
      
      # Free the C array
      LibFlow.flow_reflect_free_names(names_array, actual_count)
      
      function_names
    end
    
    # Get detailed information about a function
    #
    # @param function_name [String] Name of the function
    # @return [Hash] Function information with keys: :name, :return_type, :parameters
    #
    # @example
    #   info = mod.function_info('add')
    #   puts "Function: #{info[:name]}"
    #   puts "Returns: #{info[:return_type]}"
    #   info[:parameters].each do |param|
    #     puts "  #{param[:name]}: #{param[:type]}"
    #   end
    def function_info(function_name)
      raise FlowError, "Module is not loaded" if @handle.null?
      
      info_ptr = LibFlow.flow_reflect_get_function_info(@handle, function_name)
      raise FlowError, "Function '#{function_name}' not found in module" if info_ptr.null?
      
      begin
        info_struct = LibFlow::FlowFunctionInfo.new(info_ptr)
        
        result = {
          name: info_struct[:name].read_string,
          return_type: info_struct[:return_type].read_string,
          parameters: []
        }
        
        # Extract parameters
        param_count = info_struct[:param_count]
        if param_count > 0
          params_ptr = info_struct[:params]
          param_count.times do |i|
            param_struct = LibFlow::FlowParamInfo.new(params_ptr + i * LibFlow::FlowParamInfo.size)
            result[:parameters] << {
              name: param_struct[:name].read_string,
              type: param_struct[:type].read_string
            }
          end
        end
        
        result
      ensure
        LibFlow.flow_reflect_free_function_info(info_ptr)
      end
    end
    
    # Get a human-readable string representation of all functions in the module
    #
    # @return [String] Formatted string with all function signatures
    #
    # @example
    #   puts mod.inspect_functions
    #   # Output:
    #   # Module contains 3 function(s):
    #   #
    #   #   add(a: int, b: int) -> int
    #   #   subtract(x: int, y: int) -> int
    #   #   multiply(m: int, n: int) -> int
    def inspect_functions
      functions = list_functions
      return "Module contains no functions" if functions.empty?
      
      lines = ["Module contains #{functions.size} function(s):\n"]
      
      functions.each do |func_name|
        begin
          info = function_info(func_name)
          params_str = info[:parameters].map { |p| "#{p[:name]}: #{p[:type]}" }.join(", ")
          lines << "  #{info[:name]}(#{params_str}) -> #{info[:return_type]}"
        rescue FlowError => e
          lines << "  #{func_name} (error getting info: #{e.message})"
        end
      end
      
      lines.join("\n")
    end

    private

    # Convert Ruby value to Flow C value
    def ruby_to_flow(value)
      case value
      when Integer
        LibFlow.flow_int(value)
      when Float
        LibFlow.flow_float(value)
      when String
        LibFlow.flow_string(value)
      when TrueClass, FalseClass
        LibFlow.flow_bool(value)
      when NilClass
        LibFlow.flow_void
      else
        raise ArgumentError, "Unsupported value type: #{value.class}"
      end
    end

    # Convert Flow C value to Ruby value
    def flow_to_ruby(c_value)
      case c_value[:type]
      when :int
        LibFlow.flow_as_int(c_value)
      when :float
        LibFlow.flow_as_float(c_value)
      when :string
        LibFlow.flow_as_string(c_value)
      when :bool
        LibFlow.flow_as_bool(c_value)
      when :void
        nil
      else
        raise "Unknown Flow type: #{c_value[:type]}"
      end
    end
  end

  # Convenience methods

  # Load a Flow module from a file
  #
  # @param path [String] Path to the .flow file
  # @return [Module] Compiled module
  #
  # @example
  #   mod = Flow.load_module('math.flow')
  def self.load_module(path)
    runtime = Runtime.new
    Module.load(runtime, path)
  end

  # Compile Flow source code from a string
  #
  # @param source [String] Flow source code
  # @return [Module] Compiled module
  #
  # @example
  #   mod = Flow.compile('func add(a: int, b: int) -> int { return a + b; }')
  def self.compile(source)
    runtime = Runtime.new
    Module.compile(runtime, source)
  end

  # Get Flow language version
  #
  # @return [String] Version string
  def self.version
    VERSION
  end
end
